sequencer/src/persistence/fs.rs

```diff
@@ -1454,40 +1454,140 @@ impl MembershipPersistence for Persistence {
         l1_block: u64,
         events: Vec<(EventKey, StakeTableEvent)>,
     ) -> anyhow::Result<()> {
+        if events.is_empty() {
+            return Ok(());
+        }
+
         let mut inner = self.inner.write().await;
         let dir_path = &inner.stake_table_dir_path();
         let events_dir = dir_path.join("events");
 
         fs::create_dir_all(events_dir.clone()).context("failed to create events dir")?;
+        // Read the last l1 finalized for which events has been stored
+        let last_l1_finalized_path = events_dir.join("last_l1_finalized").with_extension("txt");
+
+        // check if the last l1 events is higher than the incoming one
+        if last_l1_finalized_path.exists() {
+            let bytes = fs::read(&last_l1_finalized_path).context("read")?;
+            let mut buf = [0; 8];
+            bytes.as_slice().read_exact(&mut buf[..8])?;
+            let stored_l1 = u64::from_le_bytes(buf);
+            if stored_l1 > l1_block {
+                tracing::debug!(?stored_l1, ?l1_block, "stored l1 is greater");
+                return Ok(());
+            }
+        }
 
-        let file_path = events_dir.with_extension("txt");
+        for (event_key, event) in events {
```

Comment:
Here we are potentially putting an unbounded large number of files into the same directory. I think this can cause perf issues, for example when listing all files in the directory. Do you know if we do that anywhere? Pruning maybe?

We're sort of deprecating file system storage but I think it's worth leaving a comment, in case we do not mange to remove it before this goes to mainnet.